pkg/store/tsdb.go

```diff
@@ -339,38 +339,47 @@ func (s *TSDBStore) LabelValues(ctx context.Context, r *storepb.LabelValuesReque
 	if err != nil {
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
-
 	if !match {
 		return &storepb.LabelValuesResponse{}, nil
 	}
 
-	if v := s.getExtLset().Get(r.Label); v != "" {
-		for _, m := range matchers {
-			if !m.Matches(v) {
-				return &storepb.LabelValuesResponse{}, nil
-			}
-		}
-		return &storepb.LabelValuesResponse{Values: []string{v}}, nil
-	}
-
 	q, err := s.db.ChunkQuerier(r.Start, r.End)
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 	defer runutil.CloseWithLogOnErr(s.logger, q, "close tsdb querier label values")
 
-	res, _, err := q.LabelValues(ctx, r.Label, matchers...)
-	if err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
-	}
+	var values []string
 
-	// Label values can come from a postings table of a memory-mapped block which can be deleted during
-	// head compaction. Since we close the block querier before we return from the function,
-	// we need to copy label values to make sure the client still has access to the data when
-	// a block is deleted.
-	values := make([]string, len(res))
-	for i := range res {
-		values[i] = strings.Clone(res[i])
+	// If we request label values for an external label while selecting an additional matcher for other label values
+	if val := s.getExtLset().Get(r.Label); val != "" {
+		if len(matchers) == 0 {
+			return &storepb.LabelValuesResponse{Values: []string{val}}, nil
+		}
+
+		hints := &storage.SelectHints{
+			Start: r.Start,
+			End:   r.End,
+			Func:  "series",
+		}
+		set := q.Select(ctx, false, hints, matchers...)
+
+		var series []labels.Labels
+		for set.Next() {
+			series = append(series, set.At().Labels())
```

Comment:
It's not needed to append all `labels.Labels` - it should be enough to `if set.Next() { return &storepb.LabelValuesResponse{Values: []string{v}}, nil }`